src/document/utils/assert-doc.js

```diff
@@ -34,4 +45,61 @@ const assertDocArray =
         }
       };
 
-export { assertDoc, assertDocArray };
+const assertDocFill =
+  process.env.NODE_ENV === "production"
+    ? noop
+    : /**
+       * @param {Doc[]} parts
+       */
+      function (parts) {
+        assertDocArray(parts);
+        if (parts.length > 1 && cleanDoc(parts.at(-1)) === "") {
+          // stripTrailingHardline can transform trailing hardline into empty string.
+          // The trailing empty string is not a problem even if it's a line element.
+          parts = parts.slice(0, -1);
+        }
+        for (const [i, doc] of parts.entries()) {
+          if (i % 2 === 0) {
+            continue;
+          }
+          if (!isValidSeparator(doc)) {
+            const type = getDocType(doc);
+            throw new Error(
+              `Unexpected non-line-break doc at ${i}. Doc type is ${type}.`,
+            );
+          }
+        }
+      };
+
+/**
+ * @param {Doc} doc
+ * @returns {boolean}
+ */
+function isValidSeparator(doc) {
+  let hasLine = false;
+  let hasUnexpectedString = false;
+  function rec(doc) {
+    switch (getDocType(doc)) {
+      case DOC_TYPE_LINE:
+        hasLine = true;
+        return;
+      case DOC_TYPE_STRING:
+        if (doc === "{' '}" || doc === '{" "}' || doc === " ") {
+          // As of now, we can include `{' '}` in line part of `fill()`.
+          // This sometimes causes overflows https://github.com/prettier/prettier/issues/2553
+          // We don't have a good way to handle this case.
+          return;
+        }
+        hasUnexpectedString = true;
+        return;
+      case DOC_TYPE_IF_BREAK:
+        traverseDoc(doc.breakContents, rec);
```

Comment:
Is this still needed if  we use the last argument in outer traverse ? 
https://github.com/prettier/prettier/blob/655eb7bdab0f3cc35b972b31c803dc7654c41db5/src/document/utils/traverse-doc.js#L24

Reply:
Maybe yes. `shouldTraverseConditoinalGroups` doesn't affect `ifBreak()`, where we want to modify the behavior.
https://github.com/prettier/prettier/blob/94fa86e1642fbec2bbb154ee8c55bc9399f21242/src/document/builders.js#L120-L132
https://github.com/prettier/prettier/blob/1ea2fd059a2ef8762f062e3540b9d48596c43b58/src/document/utils/traverse-doc.js#L68-L76
https://github.com/prettier/prettier/blob/1ea2fd059a2ef8762f062e3540b9d48596c43b58/src/document/utils/traverse-doc.js#L68-L76

Or do I misunderstand?